narwhal/worker/src/quorum_waiter.rs

```diff
@@ -117,40 +116,41 @@ impl QuorumWaiter {
-                        // A future that sends to 2/3 stake then returns. Also prints an error
+                    pipeline.push(async move {
+                        // A future that sends to 2/3 stake then returns. Also prints a warning
                         // if we terminate before we have managed to get to the full 2/3 stake.
-
+                        let mut opt_channel = Some(channel);
                         loop{
                             if let Some(stake) = wait_for_quorum.next().await {
                                 total_stake += stake;
                                 if total_stake >= threshold {
-
                                     // Notify anyone waiting for this.
-                                    if let Some(channel) = opt_channel {
-                                        if let Err(e) = channel.send(()) {
-                                            error!("Channel waiting for quorum response dropped: {:?}", e);
-                                        }
+                                    let channel = opt_channel.take().unwrap();
+                                    if let Err(e) = channel.send(()) {
+                                        warn!("Channel waiting for quorum response dropped: {:?}", e);
                                     }
                                     break
                                 }
                             } else {
-                                // TODO: maybe we should be stopping the system if that happens,
-                                //       since it seems serious. However, it may happen at the
-                                //       start of the epoch when not everyone is ready?
-                                tracing::error!("Batch dissemination ended without a quorum.");
+                                // This should not happen unless shutting down, because
+                                // `broadcast()` is supposed to keep retrying.
+                                warn!("Batch dissemination ended without a quorum. Shutting down.");
                                 break;
                             }
                         }
-                        (batch, wait_for_quorum)
+                        (batch, opt_channel, wait_for_quorum)
                     });
                 },
 
                 // Process futures in the pipeline. They complete when we have sent to >2/3
                 // of other worker by stake, but after that we still try to send to the remaining
                 // on a best effort basis.
-                Some((batch, mut remaining)) = pipeline.next() => {
-
+                Some((batch, opt_channel, mut remaining)) = pipeline.next() => {
+                    // opt_channel is not consumed only when the worker is shutting down and
+                    // broadcast fails.
+                    if opt_channel.is_some() {
+                        return;
+                    }
```

Comment:
I am ok using this in the interim, but since we want to differentiate between cases, I would suggest to make the `pipeline` futures return a `Result` , so we can explicitly handle the `Err` case instead.

Reply:
Sounds good. There is a chance that we will refactor the code more, so leaving a TODO.